docs/rfcs/022-pageserver-delete-from-s3.md

```diff
@@ -0,0 +1,290 @@
+    PS->>CP: True
+```
+
+Similar sequence applies when both local and remote marks were persisted but Control Plane still didnt receive a response.
+
+Pageserver crashed after `deleted` mark file was persisted in s3 and locally but before console received a response from PS:
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant CP as Control Plane
+    participant PS as Pageserver
+    participant S3
+
+    CP->>PS: Delete tenant
+    PS->>S3: Create deleted mark file at <br> /tenant/meta/deleted
+    PS->>PS: Create deleted mark file locally
+
+    note over PS: Crash point 3.
+    note over PS: During startup we reconcile <br> with remote and see <br> whether the mark exists
+
+    PS->>PS: delete local files other than deleted mark
+    loop Delete layers for each timeline
+        PS->>S3: delete(..)
+    end
+
+    note over CP: Eventually console should <br> retry delete request
+
+    CP->>PS: Retry delete tenant
+    PS->>CP: Not modified
+
+    note over PS: Continue with layer file deletions
+    loop Delete layers for each timeline
+        PS->>S3: delete(..)
+        CP->>PS: Finished?
+        PS->>CP: False
+    end
+
+    PS->>S3: Delete mark file
+    PS->>PS: Delete local mark file
+
+    CP->>PS: Finished?
+    PS->>CP: True
+```
+
+If pageserver crashes after both mark files are deleted then it will reply to control plane status poll request with 404 which should be treated by control plane as success.
+
+The same applies if pageserver crashes in the end, when remote mark is deleted but before local one gets deleted. In this case on restart pageserver moves forward with deletion of local mark and Control Plane will receive 404.
+
+##### Differences between tenants and timelines
+
+For timeline the sequence is the same with the following differences:
+
+- remote delete mark file can be replaced with a boolean "deleted" flag in index_part.json
+- local deletion mark is not needed, because whole tenant is kept locally so situation described in motivation for local mark is impossible
+
+##### Handle pageserver loss
+
+If pageseserver is lost then the deleted tenant should be attached to different pageserver and delete request needs to be retried against new pageserver. Then attach logic is shared with one described for pageserver restarts (local deletion mark wont be available so needs to be created).
+
+#### Summary
+
+Pros:
+
+- Storage is not dependent on control plane. Storage can be restarted even if control plane is not working.
+- Allows for easier dogfooding, console can be hosted on neon. If storage depends on control plane and control plane depends on storage we're stuck.
+- No need to share inner s3 workings with control plane. Pageserver presents api contract and S3 paths are not part of this contract.
+- No need to pass list of alive timelines to attach call. This will be solved by pageserver observing deleted flag. See
+
+Cons:
+
+- Logic is a tricky, needs good testing
+- Anything else?
```

Comment:
> I would argue that the storage does not depend on the control plane

Fair enough, what I mean here is a strict dependency. If on startup storage needs to go to control plane this is a strict dependency. For the deletion polling the worst that could happen is that data will be kept there a bit longer.

Consider the example I wrote a paragraph above:

Imagine:
PS1 fails.
PS2 gets assigned the tenant.
New branch gets created
PS1 starts up (is it possible or we just recycle it?)
PS1 is unaware of the new branch. It can either fall back to s3 ls, or ask control plane.

Currently pageserver relies on s3 ls to get list of timelines. If this source is not enough the only remaining way is to go to control plane.

> Also, albeit we did not cover the "pageserver sync" story much, I think it would be good to mention that multiple pageservers having the same tenant/timeline, may cause troubles if the deletion is called for only one of them, without altering the state of the other.
Seems like a good thing to describe anyway in the RFC.

Yep, this is out of scope, real deletion should happen only on the last pageserver that tenant is present on.

> Not a drawback, but I think worth specifying how would pageserver understand which layer files it needs to remove: if it uses index_part.json

I'm for using index part. In this case even long call to ls may not be a problem. I think combination of:
1. Delete from index_file
2. Call ls

Should be good. At step 2 prefix is supposed to be empty, so ls shouldnt take long (though would be good to test it). 

> IMO, console can do the same as long as all the "blocker" tombstones are added, forbidding pageservers to actually use such deleted entities.

It can. Though I'm a bit skeptical about letting console know about index part format as it is not specified anywhere aside from pageserver code.

Reply:
> I'm a bit confused why

I would separate cases when tenant is present continuously on several pageservers and a see-saw case when it is not present simultaneously on several nodes. Currently only see-saw case is possible, so I'm considering it in the text, but didnt want to go deeper into the second case. 

I can include a sentence like this:

In case when tenant is hosted on multiple pageservers in order to delete pageserver part of tenant data control plane needs to call detach on all pageservers except the last one, for it actual delete needs to be called. WDYT?

> As an example, aws s3 ls some/tenants/dir for 2939 tenants happens in ~5s (+ writing the results to file) for me locally.
How far were you from the s3 endpoint? There can be additional latency coming from the distance.

> The point is that console does not necessary has to know about the json file format, just be more patient instead, when listing the bucket.

Thats a good point. 
There is an option to ask pageserver to convert tenant id into a path on s3 so console can chew the data, but that anyway requires adding s3 related machinery to console (at least a corresponding library)

Reply:
> WDYT?

Yes, let's include something that mentions the situation when the tenant/timeline is present on multiple pageserver nodes, including the fact that we target the simplistic approach only here for now.

> How far were you from the s3 endpoint

I've queried a bucket from `eu-west-1`, so presumably not that far, but sure the latency is added and will be less for console.
Also, 1 list_objects_v2 returns up to 1000 keys (and then gets paginated) and one delete_objects accepts up to 1000 keys, so we don't need to list that fast 🙂

Reply:
> so we don't need to list that fast

The interesting case for me is the list of empty prefix. If we delete based on idex_part json nothing should remain there

Reply:
> let's include something that mentions the situation when the tenant/timeline is present on multiple pageserver nodes

Done